### PR TITLE
Login: Display an error on the 2FA code component when the user enters an invalid code

### PR DESCRIPTION
**Pull request by drewblaisdell**
screen shot 2017-05-03 at 4 32 47 pm

This PR updates `VerificationCodeForm` to:

- Display an error when the user enters an invalid code.
- Disable the submit button when the request is loading.
- Reset the `two_step_nonce` in memory to the new nonce returned by the API when an invalid code is entered.

**Testing**

- Visit the [`Login` page](http://calypso.localhost:3000/login) in incognito
- Submit a valid username/password for an account that has 2FA enabled and either SMS or the Authenticator app as its primary method of auth
- Submit an empty verification code
- Assert that you see an error message
- Submit an invalid verification code
- Assert that you see the error message like above
- Submit a valid verification code
- Assert that you are redirected to the Reader and logged in

- [x] Code
- [ ] Review

### CODE REVIEW

File: client/state/login/test/reducer.js [hunk 1]
```diff
@@ -28,103 +33,143 @@ describe( 'reducer', () => {
...
+			name: 'isRequestingTwoFactorAuth',
+			func: isRequestingTwoFactorAuth,
+			loadingAction: TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST,
+			failureAction: TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST_FAILURE,
+			successAction: TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST_SUCCESS,
+		},
+	];
+
+	isRequestingReducers.forEach( currentReducer => {
+		describe( `${ currentReducer.name }`, () => {
```
Comment (Tug) [inline client/state/login/test/reducer.js]: Really not a fan of generating testcases this way. We should keep tests simple and not be afraid of duplication.

**Comment by stephanethomas** [comment]
There is a problem with the server which won't return a new nonce whenever the two step code is empty or in an invalid format. Here is an example of response in that case:

```javascript
{
  "success": false,
  "data": {
    "errors": ["empty_two_step_code"]
  }
}
```

This makes the next call fails this time with the following error:

```javascript
{
  "success": false,
  "data": {
    "errors": ["empty_two_step_nonce"]
  }
}
```

This is obviously something that we should fix server-side, I'll add a card to our Trello board to take care of that.

**Comment by stephanethomas** [comment]
Following the testing instructions, I can confirm that I'm logged in and the `TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST_SUCCESS` action is dispatched.

**Comment by Tug** [comment]
LGTM 👍 
I just unrolled the loops in the test, should be good to go

**Comment by stephanethomas** [comment]
There is a few issues since the rebase, e.g.:

```
verification-code-form.jsx:117 Uncaught TypeError: (0 , _actions.loginUserWithTwoFactorVerificationCode)(...).then is not a function
  at VerificationCodeForm._this.onCodeSubmit (app:///./client/blocks/login/two-factor-authentication/verification-code-form.jsx:117:104)
  at Object.ReactErrorUtils.invokeGuardedCallback (app:///./~/react-dom/lib/ReactErrorUtils.js:70:16)
  at executeDispatch (app:///./~/react-dom/lib/EventPluginUtils.js:85:21)
  at Object.executeDispatchesInOrder (app:///./~/react-dom/lib/EventPluginUtils.js:108:5)
  at executeDispatchesAndRelease (app:///./~/react-dom/lib/EventPluginHub.js:43:22)
  at executeDispatchesAndReleaseTopLevel (app:///./~/react-dom/lib/EventPluginHub.js:54:10)
  at Array.forEach (native)
  at forEachAccumulated (app:///./~/react-dom/lib/forEachAccumulated.js:24:9)
  at Object.processEventQueue (app:///./~/react-dom/lib/EventPluginHub.js:230:7)
  at runEventQueueInBatch (app:///./~/react-dom/lib/ReactEventEmitterMixin.js:17:18)
  at Object.handleTopLevel [as _handleTopLevel] (app:///./~/react-dom/lib/ReactEventEmitterMixin.js:28:5)
```
I'm going to fix them.

File: client/state/login/actions.js [hunk 2]
```diff
@@ -13,13 +13,18 @@ import {
...
+	TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST_SUCCESS,
 } from 'state/action-types';
 
 const loginErrorMessages = {
 	empty_password: translate( 'The password field is empty.' ),
+	empty_two_step_code: translate( 'The verification code field is empty.' ),
 	empty_username: translate( 'The username field is empty.' ),
-	invalid_username: translate( 'Invalid username or password.' ),
 	incorrect_password: translate( 'Invalid username or password.' ),
+	invalid_two_step_code: translate( 'Invalid verification code.' ),
```
Comment (a8ci18n) [inline client/state/login/actions.js]: Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Invalid verification code' )` *ES Score: 19*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/2fa-error-handling)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by drewblaisdell** [comment]
Thanks, those changes look good. I added a commit to record an event when the 2FA code submission fails, which also prevents an uncaught Promise rejection. The patch has also been deployed. I'm going to check Code :+1:

**Comment by fabianapsimoes** [comment]
If I have the WordPress.com app installed for an account, the links to log in using SMS or an authenticator application don't work in this screen 

screenshot-calypso localhost 3000-2017-05-05-11-48-27

The footer links there also don't work. I suppose they need updating.

> Assert that you are redirected to the Reader and logged in

If I'm logging in from `/start` (entering an existing username in `/start/user`), I'm redirected back to signup. I think that's fine, but I'm redirected back to the beginning of the flow instead of continuing from where I stopped. 

Not sure it makes sense to fix any of those issues in this P2. Let me know otherwise, and I'll open separate issues for them.

**Comment by stephanethomas** [comment]
> If I have the WordPress.com app installed for an account, the links to log in using SMS or an authenticator application don't work
> The footer links there also don't work. I suppose they need updating.

Thanks for the review @fabianapsimoes. That's expected, it's not implemented yet.

> If I'm logging in from /start (entering an existing username in /start/user), I'm redirected back to signup. I think that's fine, but I'm redirected back to the beginning of the flow instead of continuing from where I stopped.

Good catch. Yes, it would be great to address all those issues in separate pull requests.

**Pull request merged**